### ISSUE TITLE
False positive in needless_option_as_deref if Option cannot be moved

### ISSUE THREAD

**Issue by Blaidd-Drwg** [rust-lang/rust-clippy#7846]
Lint name:
needless_option_as_deref

I tried this code:

```rust
fn main() {
    let mut i = 0;
    let mut opt_vec = vec![Some(&mut i)];
    opt_vec[0].as_deref_mut().unwrap();
}
```

Clippy complains:
```
warning: derefed type is same as origin
 --> src/main.rs:4:5
  |
4 |     opt_vec[0].as_deref_mut().unwrap();
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `opt_vec[0]`
  |
  = note: `#[warn(clippy::needless_option_as_deref)]` on by default
```

However, when applying the suggestion, the build fails:
```
error[E0507]: cannot move out of index of `Vec<Option<&mut i32>>`
 --> src/main.rs:4:5
  |
4 |     opt_vec[0].unwrap();
  |     ^^^^^^^^^^ move occurs because value has type `Option<&mut i32>`, which does not implement the `Copy` trait
  |
help: consider borrowing the `Option`'s content
  |
4 |     opt_vec[0].as_ref().unwrap();
  |               +++++++++
```
### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.58.0-nightly (bd41e09da 2021-10-18)
binary: rustc
commit-hash: bd41e09da334697c0f993b36685cb599061d9faa
commit-date: 2021-10-18
host: x86_64-unknown-linux-gnu
release: 1.58.0-nightly
LLVM version: 13.0.0
```

@rustbot label +I-suggestion-causes-error

**Comment by WaffleLapkin** [comment]
Also note that adding `as_ref`/`as_mut` is not correct in this case either, since it changes the output type to a double ref (see [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=91916b615c30ce37bec2e938872f6f93))

**Comment by H2CO3** [comment]
@WaffleLapkin this [URLO case](https://users.rust-lang.org/t/using-an-option-mut-t-in-a-loop-clippy-complains/) hits the "`as_mut()` is not enough" path as well, the fix is to `.map(BorrowMut::borrow_mut)`.
